### ISSUE TITLE
Support OAuth flow for modules

### ISSUE THREAD

**Issue by Julusian** [bitfocus/companion#2546]
### Is this a feature relevant to companion itself, and not a module?

- [X] I believe this to be a feature for companion, not a module

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the feature

Some modules use OAuth to authenticate against an API. This is done in varying ways, due to the complexities involved and it not fitting into our config model very well, as it involves navigating to a URL, and then storing a code that was returned.

The current recommended approach to OAuth is documented on the wiki: https://github.com/bitfocus/companion-module-base/wiki/OAuth

What can we do to simplify/streamline this for modules, to make the experience better for users?

### Custom config field helper

A full approach could involve creating a new config field type, which would handle the abstraction of OAuth. It would handle showing/generation of all the necessary config fields for setting up OAuth, and provide them to the module as an object.

```js
{
	type: 'oauth',
	id: 'myoauth',
	label: 'Authentication',
	width: 12,
	exchangeUrl: 'https://api.restream.io/oauth/token',
	authorizationUrl: 'https://api.restream.io/login',
	scopes: 'some scopes here',
},
```

with the resulting object looking like:
```js
{
  clientId: '',
  clientSecret: '',
  accessToken: '',
  refreshToken: '',
}
```

As part of this component we could offer a button to open the completed `authorizationUrl` so that the user can go through the process.  
We would also then handle receiving the code through a callback, and converting that into access and refresh tokens.

Some thought will need to go into the UX of this in the browser, and how that will work with saving the module config. But this would allow modules to massively simplify their OAuth usage, and make something that is consistent for the users.

Or keep it simpler with some other options to streamline some small portions:

### A way for modules to 'open' a url for the user

Some modules are using [open](https://www.npmjs.com/package/open) to open the `authorizationUrl` automatically for users. Perhaps we could provide a way for modules to request a url is opened for the user, which we can show as a prompt in the webui. That way it will work for users who are configuring remotely.

**Comment by Julusian** [comment]
This would work as after filling in the automatic fields in the settings, an 'authorize' button will become enabled.
Clicking that will open a modal hosting an iframe which performs the auth.
The callback url can point to a simple page of ours, which can emit an event to report that the auth has finished, and will push the final tokens out to the still open settings panel.
Then the user can hit save and hand over the tokens we retrieved to the module.

Maybe we should also have a 'lighter' flow for this, where we let modules create these buttons without all the config fields, and we store the whole blob of json returned in a value. That way if there is an oauth like flow that doesn't fit what we have defined, modules can still do it themselves with our help to just help collect and store the tokens

**Comment by thedist** [comment]
For my needs (utilizing multiple different OAuth flows for several different services) I think module developers themselves can handle many steps of each auth flow, but where work needs to be done is specifically on the Redirect step.

Some of the difficulties Companion faces with OAuth
1. For user-supplied credentials, they need a URL that can be redirected to. Because module devs don't know the origin for the user we often have to redirect to someone that 404's and have them copy paste the code/token.
2. For the Implicit flow modules have no easy way to access the token.
3. For dev supplied credentials, most services don't allow for wildcard redirects so we can't redirect when origin varies, and many also require HTTPS

Possible ways to solve these difficulties
1. When the redirect URL is to be done on the users side, if the the Web UI could replace a certain string with the origin being used in that web ui then the config could have something like an instruction for the user to use `$(internal:origin)/instance/${instance.label}/redirect`, where the modules own config code would format that to `$(internal:origin)/instance/SomeLabel/redirect`, and the Web UI itself would replace the origin to display `protocol://domain:port/instance/SomeLabel/redirect`
2. The OAuth Implicit flow provides a token in the URL Hash, rather than a querystring param, so to solve this there needs to be an actual page rendered and then some JS to grab the token and pass it to the module. Rather than every module doing their own version of this, a single unified Companion branded page that could be utilized by many modules would have a more professional look.
3. Having a static redirect URL used by all users of a module would allow for creating proper Companion branded apps, rather than end users needing to make dev accounts, generating credentials, and copy/pasting them in to Companion. This can't be done within Companion itself, so devs would either need to use their own server solution, or one option could be for an OAuth redirect under the `https://bitfocus.io` domain. Once redirected to that server the user could be then redirected back their Companions origin along with the token or code depending on flow for the module to handle (this can be achieved through OAuth State).

These difficulties/ideas are also ordered in difficulty to implement. The first one still requires users to make their own credentials, in some cases such as Google they'd get warnings on redirect about the app not being verified, but it would simplify at least one step of the process. The second may not be as important, as I'm not sure how many services would use the Implicit flow in this situation but for completeness it may be useful to have a solution that can be utilized across multiple modules.

The third point is one I think could be the most professional solution, as then we could have a Bitfocus Companion app for Google Sheets that all users auth to, and it could be verified so wouldn't have the warnings a user gets when using credentials they've made themselves. It would require an external site though, such as the bitfocus site, to achieve. This also doesn't solve all OAuth flows, for example Discord will always require user-submitted credentials as they refuse to verify my Companion app like they did with Elgato's Discord plugin for StreamDeck.

**Comment by Julusian** [comment]
Yes, the main difficulty is the redirect step. But starting the authorization flow is also something that we should aim to improve. 
Today there are a few approaches taken:
* Users click save, then reopen the config to retrieve the generated auth url.
* Module uses `open` after the user clicks save. This often results in broken modules on companion-pi/headless systems, only works if you are accessing companion on the local machine, and can result in browser windows opening at times you dont expect.
* Module logs the url, the user has to go find it there.
* Some combination of the above

None of these options are friendly to users.

-----

I did create https://github.com/bitfocus/companion-oauth/tree/main / https://bitfocus.github.io/companion-oauth/callback/ some months back, which was aimed to help solve this in a minimal/non-invasive manner (no core or module-api changes). But this may not work for everything, and could be improved as it is a first draft and could do with better support for multiple companion installations. Usage is described in https://github.com/bitfocus/companion-module-base/wiki/OAuth. 

-----

While I would like to explore the idea of having this as a field type, to make it really easy to do oauth, I am happy to start by implementing small chunks of this which can be used by modules to do parts of the flow manually, and then eventually also be used by this all in one field type.

So perhaps the first step could be to have a new config field type that will allow for adding an 'authorize' button in the config.
This could either call a new module method to `generateAuthorizationUrl(.....)` or perhaps it should use the expression syntax to do client side string interpolation. And another expression to check if the button should be enabled.
Then as I proposed above:
Clicking that will open a modal hosting an iframe which performs the auth.
The callback url can point to a simple page of ours, which can emit an event to report that the auth has finished, and will push the final tokens out to the still open settings panel.

It needs some thought on how to parse/fetch the token out of the iframe, considering it could be in the hash or url. (could it be under different parameter names?)
I am hoping that we would be able to use https://bitfocus.github.io/companion-oauth/something to be the redirect, and set the iframe/page to allow cross-origin to behave in an 'insecure' manner so that we can extract what we need without needing to redirect back to companion itself.

I think it is worth at least prototyping this to see how it would work in practise (from a UX perspective)

-----

> Having a static redirect URL used by all users of a module would allow for creating proper Companion branded apps, rather than end users needing to make dev accounts

I think this would be good, but we should be careful to ensure that modules wont become unusable because the tokens are no longer valid. Because that could happen at any time, including on a show day for someone. And if they are using an older release, being forced to update companion to resolve this will be incredibly annoying or unrealistic as it could require them to rebuild chunks of their config

**Comment by thedist** [comment]
First off I think an iframe should be ruled out, as while the OAuth 2.0 spec doesn't strictly prohibit the use of iframes it strongly encourages OAuth providers to use `X-Frame-Options` to deny it, which is why most OAuth providers (Twitch and Google to name a couple off the top of my head) block auth pages from being loaded inside an iframe.

> It needs some thought on how to parse/fetch the token out of the iframe, considering it could be in the hash or url. (could it be under different parameter names?)

If a service is compliant with the specification then the implicit flow would redirect with an `access_token` in the URL hash, and for the Auth Code flow it'd be a `code` in the querystring params.

> I think this would be good, but we should be careful to ensure that modules wont become unusable because the tokens are no longer valid. Because that could happen at any time, including on a show day for someone. And if they are using an older release, being forced to update companion to resolve this will be incredibly annoying or unrealistic as it could require them to rebuild chunks of their config

I don't see anything that would require a user to update their Companion to get new tokens or anything like that. OAuth tokens expire and need refreshing periodically (for example, every hour for Google, 4 hours for Twitch), if a user provides their own credentials then everything could be done in the module to make requests to the service to get new tokens, but that means they have to make their own apps, enable the APIs they need, copy/paste credentials etc... With my suggestion it'd be the module devs making a single app and the credentials and token renewal would be handled outside of the module, with just the access token being passed to the module. This would mean that an outage of that service running outside of Companion would break things, but that has to be weighed against the somewhat amateur nature of the way we're doing things currently needing users to do dev stuff.

One potential idea that just came to mind:
1. User goes through a modules config, entering whatever they need, then click save.
2. If the module has everything it needs for OAuth (such as credentials), it goes to a new 'Awaiting Auth' state. This state would show on the Connections screen (which the user would be on after saving the config), and could be clickable and send the user to the service. As part of the generation of the URL, we could encode the module instance id, and origin, into the `state` param.
3. The user would agree to the auth, and be redirected to some place like `https://bitfocus.github.io/companion-oauth/callback/`. This would be HTTPS, as some providers require that, that site could then decode the `state` param to know where to redirect the user to, and so redirects the user to that URL (which is their Companion web ui), to some path that Companion could use for auth where it takes the params from the URL, passes them to the module (of which it has the instance ID from the state param). and then auto-close the tab.
4. The auth tab should now be closed, the module has the tokens, and can switch to a Connected state.

**Comment by Julusian** [comment]
> First off I think an iframe should be ruled out, as while the OAuth 2.0 spec doesn't strictly prohibit the use of iframes it strongly encourages OAuth providers to use X-Frame-Options to deny it

Good to know, I wasn't aware of that. I agree that this shouldn't be an iframe then.

> I don't see anything that would require a user to update their Companion to get new tokens or anything like that. OAuth tokens expire and need refreshing periodically

Sorry, I meant the client id and client secret. 

> As part of the generation of the URL, we could encode the module instance id, and origin, into the state param.

Ooh, good idea. For some reason it didn't occur to me to put the current origin into there..  
We would also want to include the config field name(s) to write the result into, probably some kind of version number so that we can change the format without changing the url too. But we don't have to define this too much right now, as its not a 'public api' even though it is called externally.

> When the token expires, if the user supplied their own app credentials then refreshing a token could be done in module, otherwise it could make the request to an external server run by the module dev or bitfocus to handle token renewal with the credentials stored outside the module.

For now lets leave this as up to the module to figure out.  
I'm open to this, but have questions/concerns about scalability (will some providers see 1000s of different client-ids being used from the same ip as spam/bots and revoke tokens or ban accounts? It sounds kinda suspicious, perhaps they could be stolen credentials that are being collected/kept for an attack), or could it simply be against TOS? 
It definitely ought to be possible to opt-out of this auto-renew service (or should it be opt-in), some places/people will not like their tokens being held by an external service like this.

>  This state would show on the Connections screen (which the user would be on after saving the config), and could be clickable and send the user to the service

This will need a bit of thought on how/where to generate the url and get it to the ui (should the second parameter be an object for this state? should it be driven by the manifest.json? something else?).  
I have had the thought about maybe we could show a notification that a module needs oauth completing, which this would allow for nicely.

So this approach sounds good to me :)

**Comment by jswalden** [comment]
I might try my hand at implementing this.

(Exactly precisely what "this" is, in fine-grained API detail, is likely to remain vague until I've got something hacked together.  Once _something_ is working, it should be easy enough to iterate on it.)

**Comment by jswalden** [comment]
Okay, I implemented something.

Build this companion-module-base: https://github.com/jswalden/companion-module-base/tree/oauth-support  (or https://github.com/jswalden/companion-module-base/commit/5f3646ca772111d68854a6b43629ac15af128919 for the exact commit in case I go back and change anything more)

Then pull this branch, somehow edit the `companion/package.json` to forcibly link against/use the preceding c-m-b or use `yarn link` or whatever, and build it: https://github.com/jswalden/companion/tree/native-oauth-support  (or https://github.com/jswalden/companion/commit/1c455add202280de29e16459c9cbcf1e9d7c7aed for the exact commit in case I go back and change anything more)

The end result is unpretty as UI because that wasn't the point.  The internals contain `XXX` comments all over the place for various stuff like implementing authorization timeouts or window closure or whatever.  🙂  But it all does in fact give a way for a module to insert a config-field that looks like, say, this:

```js
		{
			type: 'oauth',
			id: 'oauth-info',
			width: 12,
			label: 'Google application OAuth parameters',
			authorizeEndpoint: `https://accounts.google.com/o/oauth2/v2/auth?prompt=consent&access_type=offline`,
			tokenEndpoint: 'https://oauth2.googleapis.com/token',
			scopes: ['https://www.googleapis.com/auth/youtube.force-ssl'],
		},
```

and then the only thing you have to enter is the client ID/secret, and maybe set the port the OAuth callback server runs on -- it's hacked in as a service, and host/port are set from a userconfig setting, and the widget will display the redirect you need to add to the service's authorized redirect list -- and hit the Authorize button.  An authorization URL will be constructed using your chosen client ID/secret, incorporating the scopes specified by module developer, with a `state` parameter that contains an anti-CSRF nonce plus the origin of the Companion UI, and then it's opened in a popup.  Upon successful completion of the auth process, the OAuth resource redirects back to the Companion OAuth callback which will validate and process the response parameters (client-side, solely because I didn't want to deal with safe interpolation of user-generated strings into HTML), then `postMessage` authorization results back to the Authorize button's onClick handler to continue and fetch actual tokens.  This all predated -- and still predates -- introduction of secrets, so I just shoved all of it into the option value as a new type in `InputValue`.

I did not actually add some kind of service to refresh retrieved access tokens in the background, but that would be an obvious thing to do.  Some thought would have to be given to how to deal with failed refreshes.  (I'm definitely aware that e.g. churches using this for YouTube would probably  get regularly bitten by Google's one-week token expiration period if they don't have Companion constantly running, just for one ready example.)  It would be pretty plumby, not technically complicated.

I didn't actually bother making my hacked-up `youtube-live` module internally access the tokens this all would create, or deal with their non-acquisition.  But they all show up in a connections export JSON so they're presumably there to use.

![Image](https://github.com/user-attachments/assets/6d46540f-8a75-4d68-bc00-08f3cd01e510)

(For a time I thought I needed #3494 in order to be able to redirect back to the specific Companion connection tab active, but that was just me doing a dumb thinko in how I compared message origin against OAuth callback server origin in the message-receiving handler.  The approach could be changed to not use a popup and to redirect back to the edit tab, with some trouble.)

Lots of unprettiness in all of it, but that wasn't really the point.  🙂

I'm very unsure where one would want to go with all this at this point.  If you're accessing the Companion UI on the selfsame machine, this all works.  But the fundamental problem of needing a stable scheme://host:port that works _for the current user on his machine_, and for all users accessing the Companion web UI from their machines, remains.

Oh, and also some services surely require a TLS OAuth redirect URI if it's not on the selfsame machine.  (Google allows HTTP for localhost/127.0.0.1, but I don't know if that extends to all private network address ranges or not.)

I hope we can agree that making a Bitfocus endpoint for all this that runs on the public web is security/privacy-awful for most users.  Convenient, sure.  But basically it's saying Bitfocus gets everyone's access and refresh tokens to do anything any user does, which does not seem acceptable at all and which I can't imagine Bitfocus wanting to incur such liability anyway.

So, yeah.  I'm more or less stuck at this point as far as ironing out the actually hard questions here.

**Comment by Julusian** [comment]
Sorry, I've neglected this.

>  it's hacked in as a service

Is there a reason for this rather than using the existing http server?  
I suppose there would need to be a class to handle some of the refresh portions of this, but maybe not as a full 'service'.

> I did not actually add some kind of service to refresh retrieved access tokens in the background, but that would be an obvious thing to do. Some thought would have to be given to how to deal with failed refreshes.

I think this should be done before this gets merged, but this definitely doesnt need to happen at this POC stage.  
I suspect we simply shouldn't do anything with failed refreshes other than log something, and maybe retry it a few times?
Either way, modules will need to handle the situation of a bad token, and prompting the user to reauth can be something else to tackle (some icons in the header indicating things that need addressing, such as any modules in an 'auth failed' state. for oauth that could give links to reauth, for others a link to the module config?)

> regularly bitten by Google's one-week token expiration period

I would definitely say lets start with a local reauth ability, and thinking about what we can/should do for those users as a follow up task.

> I'm very unsure where one would want to go with all this at this point. If you're accessing the Companion UI on the selfsame machine, this all works. But the fundamental problem of needing a stable scheme://host:port that works for the current user on his machine, and for all users accessing the Companion web UI from their machines, remains.

I think there are 2 cases to consider here;
1) companion is run on a server and is always a predictable hostname
2) everyone else

For 1, that seems easy, we just need a way to determine that url which could be something in the userconfig. (maybe with some fancy ui to prompt to configure this on first use of oauth. as a follow up)  
For 2, I would lean into using a tiny github hosted endpoint to do the redirecting. I made https://bitfocus.github.io/companion-oauth/callback/ a while back for this purpose, modelled on what home-assistant offers. If we can fully automate it based on a state value then even better.  (feel free to have a rummage [here](https://github.com/bitfocus/companion-oauth), I don't think any module uses it yet, so it can be heavily reworked as needed)
By using github pages, we are potentially leaking the values to github, but not to bitfocus (which may be better or worse, depending on who you ask). And as part of that fancy ui to configure oauth behaviour, we can make using this opt-in, so those who don't like it can do it via option 1, or the painful way.  

> Oh, and also some services surely require a TLS OAuth redirect URI 

The github pages redirector mitigates this issue too.

Maybe we can limit the potential leak of codes by pushing modules to use a fragment based token redirect url instead of query based whenever possible? It looks to be a supported thing for some providers https://openid.net/specs/oauth-v2-multiple-response-types-1_0.html#ResponseModes

**Comment by thedist** [comment]
I wish more services supported the Device Code Flow for OAuth, it's designed exactly for Companions use case and has simplified things significantly for my Twitch module (until the routing changes in Companion broke the link).

In terms of other auth flows though, a fragment based token (the Implicit flow) is a non-starter for me, by design it has no capability of refreshing since it's designed for things like web apps, as each time the user visits the site they can be run through an OAuth flow transparent to the user and have a fresh token. This doesn't work for Companion though, not everyone has the Web UI up, and there's no easy way to make authentication with the Implicit flow transparent to the user. Worst case scenario, a token does during live production, and it remains dead until user intervention. Oh, and also being a fragment means it's not accessible to the Companion web server, you'd need some client-side JS in the Web UI to grab the token and pass it back to Companion so that the module can actually use it.

For my usages, such as the Google Sheets module, already handle all of the code exchange and refreshing process entirely within the module itself, the only janky bit of it is the redirect URL and the user needing to copy the code from the URL and paste it into the config, but this could easily be handled by having a redirect url point to a github page, or a bitfocus page, which all it does is use a little JS to render a nice page saying 'copy this code, and past it into the Config'. It could even potentially just redirect straight to Companion but that would require something in the Web UI to encode the Companion ip/domain and port into the state param of the OAuth URL so that the github page/bitfocus page could redirect the user on to Companion.

**Comment by jswalden** [comment]
> > it's hacked in as a service
> 
> Is there a reason for this rather than using the existing http server? I suppose there would need to be a class to handle some of the refresh portions of this, but maybe not as a full 'service'.

A large part is that this is Security, and for Security making the server as absolutely tightly scoped and isolated as possible seemed ideal to me.  You can _look_ at this new server and see how it's locked down, there's only one thing on it.  Whereas the main server has all sorts of stuff on it.

I also think I may have run into complexities about figuring out how to shove this one bespoke endpoint into the existing HTTP server.  Or maybe not.  Memory hazy, I was proof-of-concepting hard as I came to understand OAuth better and better.

> I think this should be done before this gets merged

For clarity I do not expect this to be merged any time soon and don't think it's terribly near to it being mergeable.  🙂 

> For 2, I would lean into using a tiny github hosted endpoint to do the redirecting. [...] By using github pages, we are potentially leaking the values to github, but not to bitfocus (which may be better or worse, depending on who you ask). And as part of that fancy ui to configure oauth behaviour, we can make using this opt-in, so those who don't like it can do it via option 1, or the painful way.

Yeah, I dunno about this.  I don't think Github wants to have people's OAuth access/refresh tokens any more than Bitfocus does.  🙂  Nor I would hope would most users -- but I know enough to know that most would just smash that `[Whatever]` button if you offered it without actually understanding the consequences.  Which does not seem great!  I would feel guilty about offering users ways to incur risk out of nothing but convenience.

> Maybe we can limit the potential leak of codes by pushing modules to use a fragment based token redirect url instead of query based whenever possible? It looks to be a supported thing for some providers https://openid.net/specs/oauth-v2-multiple-response-types-1_0.html#ResponseModes

To me it feels like if you're gonna do this, you really should just knock out all OAuth services at once, which means use the most-supported thing and not use more esoteric or disfavored flows.

My gut keeps telling the only real solution, is for Companion to grow the concept of having a canonical and user-definable IP/hostname of access (possibly requiring them to tailor their local DNS environment to support this) that is TLS-accessible, and redirect non-canonical uses to it.  Then you have a stable, *user controlled and local* URI you can use for redirects.  Everything else just makes serious security compromises that users will not think through before accepting.

**Comment by jswalden** [comment]
> This doesn't work for Companion though, not everyone has the Web UI up, and there's no easy way to make authentication with the Implicit flow transparent to the user.

Yeah, Companion is kind of akin, in a weird way, to a limited-input device as far as all this goes.

> It could even potentially just redirect straight to Companion but that would require something in the Web UI to encode the Companion ip/domain and port into the state param of the OAuth URL so that the github page/bitfocus page could redirect the user on to Companion.

My li'l hack made state into a nonce-webUIOrigin string tuple to perform just such a redirect (or rather to `postMessage` to the opening window, but it's kind of all the same really).

**Comment by thedist** [comment]
One thing to also remember for OAuth, is that some services (such as Google) have verification and auditing requirements for apps that require sensitive or restricted scopes. Google Sheets for example uses Sensitive scopes, which would require verification. Companion would likely fail this, and it doesn't meet the requirements for passing the security audits for any module that may need restricted scopes.

Because of this I'd recommend the user providing their own app credentials rather than having Bitfocus managed apps with credentials having to be stored somewhere. This way Bitfocus/module devs don't have to worry about storing credentials to be used by all users, or figuring out if the code exchange/refresh process should happen within Companion, a Github page, or somewhere else, as it can all happen within the module itself with the user provided credentials.

The only real thing that needs tackling IMO is the redirect portion of the OAuth process, which could be handled:
1. User provides Client ID, and Client Secret in the config of a module.
2. Config page tells users to set a Redirect URI in their app for `https://some.static.page/by/bitfocus`
3. User clicks save on the config, then goes to the modules HTTP handler for /instance/NAME/auth, and passes the origin as a param (this needs to be done within the Web UI, as the module wont know the origin).
4. The modules HTTP handler generates a nonce, any metadata needed, combines them with the origin (and appropriate path) and encodes into the `state` param, then redirects the user to the OAuth page for whatever service it's for.
5. User completes the OAuth accept/deny step and is redirected to `https://some.static.page/by/bitfocus`, that page decodes the state param to get the origin, and redirects the user to Companion, along with the `code` param and `state`.
6. the HTTP handler within the module can perform any checks against the nonce, exchange the `code` for an access_token and refresh_token, and periodically refresh as needed without further user intervention.

No need for the app credentials to be stored anywhere but the users Companion system, the redirect URL can be a very simple dumb site that just redirects back to Companion but ticks the HTTPS requirement, and with the WebUI passing the origin to the modules auth handler it'll work if the user is running on an ip:port, or has Companion behind a VPN with PAT, or nginx, or anything like that as the origin will be whatever address the user is accessing in their browser.

**Comment by Julusian** [comment]
> In terms of other auth flows though, a fragment based token (the Implicit flow) is a non-starter for me, by design it has no capability of refreshing

I didnt realise that (I have a feeling it was mentioned somewhere above), so that is out then

> Oh, and also being a fragment means it's not accessible to the Companion web server, you'd need some client-side JS in the Web UI to grab the token and pass it back to Companion so that the module can actually use it.

I don't see that as a problem. We can easily do this client side, and it might even be better to.

> For my usages, such as the Google Sheets module, already handle all of the code exchange and refreshing process entirely within the module itself,

Cool, but the aim of this is to make it easy for modules to give users a nice flow as much as possible. I will advocate for modules moving over to it, but it wont be a hard requirement.   
Once we have an easy flow defined, I am open to seeing what portions we can expose separately for modules which want to do some things their own way.


---

> A large part is that this is Security, and for Security making the server as absolutely tightly scoped and isolated as possible seemed ideal to me. You can look at this new server and see how it's locked down, there's only one thing on it. Whereas the main server has all sorts of stuff on it.

I guess so, but unless you audit the flow of module config throughout the rest of companion, is this adding anything? 

For me, this might make it less secure as I run companion behind a nginx proxy with ssl, and then setting up this oauth is something to be the https is another step that I wouldnt realise had to be done initially, so would be likely to skip it for at least the first time.

> Yeah, I dunno about this. I don't think Github wants to have people's OAuth access/refresh tokens any more than Bitfocus does. 

No, but my thinking here is that github probably has better log security/rotation processes than bitfocus will, so it seems like a safer bet in most cases.  
Something else to question is does this actually open them up to more risk? Surely the code returned is one-use (and short-lived?), so as soon as companion converts it into the real token the code will be useless?

> To me it feels like if you're gonna do this, you really should just knock out all OAuth services at once, which means use the most-supported thing and not use more esoteric or disfavored flows.

I meant we should default to the fragment flow, but allow modules to opt into the query flow if needed. not to limit what we support.

> possibly requiring them to tailor their local DNS environment to support this
> Then you have a stable, user controlled and local URI you can use for redirects.

I think that will be way too much for many users. I could refer to the many times that atem module users havent known how to set a static ip address, or give both their machine and the atem the same ip address.  
Making a requirement like this would break many peoples environments too, 

----

> Because of this I'd recommend the user providing their own app credentials rather than having Bitfocus managed apps with credentials having to be stored somewhere. This way Bitfocus/module devs don't have to worry about storing credentials to be used by all users, or figuring out if the code exchange/refresh process should happen within Companion, a Github page, or somewhere else, as it can all happen within the module itself with the user provided credentials.
>
> The only real thing that needs tackling IMO is the redirect portion of the OAuth process, which could be handled:

That sounds like almost exactly what I am asking for other than I want companion to have a concept of ouath so that it can do more of the heavy lifting and cut down on the amount of boilerplate modules have to do.

Implementing that http handler in a module is doable, but its basically just copy and paste. And don't forget the 3+ fields needed to make this work, maybe should have a code input field just in case they can't/won't use a redirect handler. And a way to clear any old tokens. All adding to the boilerplate (or in reality, many modules just won't do most of that as it is nicities). If we have a config field type for oauth which builtin handles 90% of this then that makes it easy for modules.
 
And because this is all in the connection config panel, we should probably think about ux as the user will need to save the config before the link will work, and that closes the panel, so some work probably needs to go into this.

**Comment by jswalden** [comment]
> > A large part is that this is Security, and for Security making the server as absolutely tightly scoped and isolated as possible seemed ideal to me. You can look at this new server and see how it's locked down, there's only one thing on it. Whereas the main server has all sorts of stuff on it.
> 
> I guess so, but unless you audit the flow of module config throughout the rest of companion, is this adding anything?

You're right that it's technically not, as all of a Companion instance is same-origin so basically the whole instance has access to all stuff at once.

There is some small value in that fixing the OAuth host:port this way allows the web UI host:port to be changed without having to register a new redirect URI with the OAuth-supporting service.  (You could argue this is too small a value if you wanted.)

> For me, this might make it less secure as I run companion behind a nginx proxy with ssl, and then setting up this oauth is something to be the https is another step that I wouldnt realise had to be done initially, so would be likely to skip it for at least the first time.

Unfortunately, the hard thing I've learned over time is that security just frequently does not integrate seamlessly into the rest of otherwise-elegant setups.  It messes stuff up, it breaks previous assumptions, it turns ignorable implementation details into vulnerabilities, it can often make life non-marginally harder.  The activation energy to get a proper, fixed-address TLS endpoint up is simply a non-marginal increase in pain secure OAuth demands.  😐 🤷

> Surely the code returned is one-use (and short-lived?), so as soon as companion converts it into the real token the code will be useless?

Authentication codes are _supposed_ to be one-time use, yes.  (And multiple uses is _supposed_ to make the authorization server invalidate existing access/refresh tokens acquired using the authentication code.)  But if the auth server doesn't implement that, or if it doesn't implement the PKCE extension (which my changes dutifully supplied, but it's entirely up to the authorize/token endpoints to actually enforce) and if the attacker _somehow_ can prevent Companion from trying to use the authentication code to acquire access/refresh tokens, then the authentication code (and state) are as good as an access token.

Functionally, the attack we're worried about here where "Github" (or a malicious employee) steals the authentication code is identical to the one described in [RFC 7636 §1](https://datatracker.ietf.org/doc/html/rfc7636#section-1).  The PKCE extensions bits I added (`code_challenge`/`code_challenge_method` in authorization request, `code_verifier` in token request) would prevent that attack -- but only if the particular OAuth server correctly implements RFC 7636.

> I think that will be way too much for many users.

I agree it's hard.  But then if they're running Companion on a non-loopback and intending multiple machines to access it using that non-loopback address, then like it or not, functionally they've become the administrators of a web server.  It's possible the hard comes with the terrain, unfortunately.

> Implementing that http handler in a module is doable, but its basically just copy and paste. And don't forget the 3+ fields needed to make this work, maybe should have a code input field just in case they can't/won't use a redirect handler. And a way to clear any old tokens. All adding to the boilerplate (or in reality, many modules just won't do most of that as it is nicities). If we have a config field type for oauth which builtin handles 90% of this then that makes it easy for modules.
> 
> And because this is all in the connection config panel, we should probably think about ux as the user will need to save the config before the link will work, and that closes the panel, so some work probably needs to go into this.

To be honest, I'm not at all sure that exposing these values via config object is the right developer UI.  Especially not when the access token should be considered a "secret" and the refresh token arguably ought not be exposed at all (and that some servers won't even return at all).  I tried to think briefly about what the right UI would be when writing the patch, but then I remembered I was proof-of-concepting and this was the wrong time to think hard about it.

I think the right interface would be something that lets module code ask for an access token, be informed if the access token is invalid and refreshing failed and that the user must reauthorize, and possibly have some way to wait for a reauthorization to succeed/fail in order to retry.  Exposing this on the config (or in secrets) limits the module code to doing nothing other than very crude polling.

**Comment by Julusian** [comment]
> You could argue this is too small a value if you wanted

Yeah I am not convinced of the value of this or if it is going to be any easier. I don't know why anyone would change the companion port/address after setting it up and not also have to change the oauth one.

>  then like it or not, functionally they've become the administrators of a web server. It's possible the hard comes with the terrain, unfortunately.

I think that depends on the setup. A fair chunk of the userbase is temporary (1-2 week long) installs, who definitely wont to have to make dns changes (they may not even be able to) or setup anything more than self-signed ssl 

> Authentication codes are supposed to be one-time use, yes
>  The PKCE extensions bits I added would prevent that attack 

Then this is sounding more and more like there is little risk in using a public redirect server then to me.   


> To be honest, I'm not at all sure that exposing these values via config object is the right developer UI. Especially not when the access token should be considered a "secret" and the refresh token arguably ought not be exposed at all (and that some servers won't even return at all). I tried to think briefly about what the right UI would be when writing the patch, but then I remembered I was proof-of-concepting and this was the wrong time to think hard about it.

Well this should definitely be forced to use the new secrets object that sits parallel to the config, which means that once saved these values arent sent back to the ui (I think they are kept out of the exports too), so are kept secret as best as we can.
But of course modules will need to be given their access tokens. Refresh token will depend on whether we do the refreshes for them or leave that for them to handle.

I am envisioning the ui being done so that the oauth 'component' shows as a line of text and an icon, saying something like: 'auth not setup', 'authorization needed' , 'token expires 11:11:11 12/23/45 (missing refresh token)' or 'auth ok'. With an edit button next to it that will either expand this to reveal more fields, or will do so in a modal.  
In that, it will be possible to set/update any values (using the secrets approach of today, just indicating whether values are set) and then saved.  
Somewhere here will also be a button that opens the auth window, and when complete fills the code into the right slot. 

> Exposing this on the config (or in secrets) limits the module code to doing nothing other than very crude polling.

surely your proposal means they have to poll as companion wont be able to tell them that the token has been refreshed?  
By putting it in the existing secrets object, we can simply call `updateConfig` whenever the secret changes and assume the module will handle that sensibly. I am hesistant to introduce a whole new flow for these values, that will counter this effort and will introduce some new boilerplate/repetition.  
I suppose one question here is how should the module know whether the token was updated because it was refreshed or if it updated because it is a different user. But we could indicate that in some way

**Comment by jswalden** [comment]
> A fair chunk of the userbase is temporary (1-2 week long) installs, who definitely wont to have to make dns changes (they may not even be able to) or setup anything more than self-signed ssl

Self-signed SSL is the minimum for this, I think.  Something like a usable DNS name is more gravy than anything, yet might be a nice-to-support for longer-lived setups that customize the canonical-host setting.

(1-2wk is a fascinatingly small duration, I wouldn't have expected so short, honestly.  I'm idly curious why it would be that short -- is this really people not having their own wifi router to run all their stuff on their own private subnet, NAT-isolated from some premises's network? only reason I can think of, off top of head)

> > Authentication codes are supposed to be one-time use, yes
> > The PKCE extensions bits I added would prevent that attack
> 
> Then this is sounding more and more like there is little risk in using a public redirect server then to me.

With any service supporting and requiring PKCE, at least.  Even so, is this sending off to Github going to pass the Google review gauntlet, or similar?  I'd bet against it.

> surely your proposal means they have to poll as companion wont be able to tell them that the token has been refreshed?

Refresh tokens should be largely invisible to application logic, with only managerial code (inside Companion) handling their use when needed.  If you mean access token here, then whatever `InstanceBase`-homed function (presumably taking the config option id as argument) that would expose this would return the instantaneously valid access token -- and we would simply encourage people to freshly call that function every time they need to use their access token.

Actually tho, now that I think of it -- maybe the right way is something like (or report failures as typed return values, or whatever)

```js
class InstanceBase {
  async performOAuthOperation(
    oauthOptId: string,
    act: (accessToken: string): Promise<boolean>, // whether the OAuth resource says the access token is valid, orthogonal to operation success/failure
  ): Promise<void> {
    if (not authenticated at all for oauthOptId service)
      throw new Error("not logged in");

    // have access token for oauthOptId service
    if (act(saved access token for oauthOptId service))
      return;

    if (!have refresh token for oauthOptId service)
      throw new Error("don't have refresh token to acquire access token, please relogin");

    try exchanging refresh token for new access/refresh tokens for oauthOptId service;

    act(new retrieved access token for oauthOptId service);
  }
);
```

and then the module developer would pass a function that performs the operation, that receives the access token within and (hopefully -- hammer on this in docs and at that point hope for the best) uses it for the bounds of that function frame only.  You could also use some return sentinel value to indicate needing to get a new access token -- that would let you combine the two functions into one.

> I suppose one question here is how should the module know whether the token was updated because it was refreshed or if it updated because it is a different user. But we could indicate that in some way

At end of day, a module's business logic just wants the opaque access token string it can use to do the stuff it wants to do.  Everything else -- any refresh token, the client ID/secret, OAuth server endpoint URLs, scopes, all the rest -- is only incidentally needed to get one of those, or updating to a new one, etc. I think.

**Comment by Julusian** [comment]
> (1-2wk is a fascinatingly small duration, I wouldn't have expected so short, honestly. I'm idly curious why it would be that short -- is this really people not having their own wifi router to run all their stuff on their own private subnet, NAT-isolated from some premises's network? only reason I can think of, off top of head)

Well the original target audience is live av work. So 1 day shows (plus setup and rehearsals) or maybe a few days of show are most common there.  
For me personally, I use companion for things around my house and office, and then on shows/productions lasting at most a week (perhaps with a config partially pre-setup a few days in advance). And some of these would definitely want to use google-sheets, not sure about anything else using oauth

> With any service supporting and requiring PKCE, at least. Even so, is this sending off to Github going to pass the Google review gauntlet, or similar? I'd bet against it.

I'm not proposing making this the only option, but an easy and suggested option. As I have it setup under a proper dns name in my setup I would want to do the callback directly. But when doing a show having that callback service could be much easier.  
All that we are discussing here is some defaults/presets to offer here when a user first encounters this oauth config field, it will always want to be able to accept any url/hostname they want.

> Refresh tokens should be largely invisible to application logic, with only managerial code (inside Companion) handling their use when needed.

I agree with that (unless we let a module opt into manual token refreshes for some reason)

> and then the module developer would pass a function that performs the operation, that receives the access token within

I'm not sure about that approach. For some modules that will be painful.  
I don't have a good example of this, but if a module is using a library for interacting with an api, that library will likely expect to be given the authtoken once and then reuse it for multiple calls. It looks like youtube-live and rather old versions of spotify-remote would fall into this.

> At end of day, a module's business logic just wants the opaque access token string it can use to do the stuff it wants to do. 

Sure, but when the logged in user changes the module will probably need to do some scraping of data, or invalidate some caches. When the token gets refreshed it wont need to do those things, of course it could for simplicity but that might have too much cost for some modules

**Comment by jswalden** [comment]
> Well the original target audience is live av work. So 1 day shows (plus setup and rehearsals) or maybe a few days of show are most common there.

I'd just have figured you'd have your own router with its own subnet set up, so everything within that particular IP range stays pretty much the same all the time -- only the router's WAN would change site to site.  But we digress.  🙂

> I agree with that (unless we let a module opt into manual token refreshes for some reason)

Seems reasonable to support manually refreshing, yeah.

> I'm not sure about that approach. For some modules that will be painful. I don't have a good example of this, but if a module is using a library for interacting with an api, that library will likely expect to be given the authtoken once and then reuse it for multiple calls. It looks like youtube-live and rather old versions of spotify-remote would fall into this.

Hmm.

> but when the logged in user changes the module will probably need to do some scraping of data, or invalidate some caches. When the token gets refreshed it wont need to do those things, of course it could for simplicity but that might have too much cost for some modules

That seems to argue for exposing *when* a full authorization is triggered, that is rarer than the authentication code itself changing.  Keying on the contents of the access token, which is all that relying purely on `configUpdated` allows, is keying on the wrong thing.
